### DISCUSSION TITLE
Kustomize and Helm diff support

### DISCUSSION THREAD

**Discussion by willejs** [fluxcd/flux2#820] [Proposals]
Usually, one really valuable step in your CI against a PR is to do a kubectl/helm diff to show users what would change in your cluster (minus secrets). 

A great example of this is probably how Terraform Enterprise/Spacelift/Atlantis do this.
They will run a terraform plan and post the diff of the plan back to your PR as a comment and link through to the plan phase for you to inspect. This is really helpful when reviewing PRs, catching errors early on, and communicating the impact of your PRs. In the CiOps world, it is common for people to use helmfile to do a similar thing too.

Currently, you could probably hack this into your CI, reading a helmRelease/kustomize resource and trying to construct diff commands. This would end up devaluing one of the main benefits of gitops, not having k8s credentials in CI tasks. Additionally, it would also mean you have to try and replicate the same logic that the helm (merging values) and kustomize (sops secrets), and even source (repackaging helm values files) controllers carry out.

I am wondering if there is a way to implement this in flux2. Getting the source controller to create an artifact for a PR, letting the controllers do the diff, send a notification to a github pr, and store the diff output in a resource, cleaned up on PR close? Just an idea. 

Really, I am wondering if this is on anyone else's radar, and If this could become a feature request?

**Comment by stefanprodan** [comment]
Posting the diff of decrypted data(Kubernetes secrets) to GitHub or any other external system it’s a no-go IMO.

**Reply by william-salt-cko** [reply]
Totally agree, most existing tools, [terraform](https://github.com/runatlantis/atlantis/issues/163) and [helm diff](https://github.com/databus23/helm-diff) do not output secrets in diffs when used correctly. Kubectl has [an open issue](https://github.com/kubernetes/kubernetes/issues/87840) to implement this.

**Reply by stefanprodan** [reply]
Ok we can revisit this feature request for kustomize-controller when kubectl adds support for hiding secrets.

**Reply by william-salt-cko** [reply]
I am wondering if its worth visiting still for helm users? Do you know your user base split between people using helm vs kustomize? I would presume most people are using helm and would appreciate the feature, implementing it for kuztomise would be easier later on too?

**Reply by stefanprodan** [reply]
It's kustomize-controller that applies HelmReleases and HelmRepositories on the cluster, so you can't get a diff of those from helm-controller. As for the release values, we have the same issue with secrets, as [values can come from Kubernetes secret](https://toolkit.fluxcd.io/guides/helmreleases/#refer-to-values-in-configmap-and-secret-resources), if we would post the helm diff to an external system it will contain the secret data mapped to values.

**Reply by william-salt-cko** [reply]
Kustomize controller aside. I think if it only posted the output from the diff from the `helm diff` command it would be fine, as helm diff suppresses secrets.

**Reply by hiddeco** [reply]
Suppressing secrets does not mean that there is no sensitive data in the other resources.

**Reply by stefanprodan** [reply]
@william-salt-cko the [valuesFrom](https://toolkit.fluxcd.io/guides/helmreleases/#refer-to-values-in-configmap-and-secret-resources) maps Kubernetes secrets to Helm values. How would helm diff know to hide a value like `aws.token: <TOKEN-TAKEN-FROM-SECRET>`.

**Reply by william-salt-cko** [reply]
`aws.token: <TOKEN-TAKEN-FROM-SECRET>`  Would go into a secret resource created by the helm chart, which would be suppressed by the `--supress-secrets` flag in helm diff. It would not get leaked.

If someone chooses to put a secret string via a value into any other resource that is not a secret, it will indeed diff and leak it. However, secrets shouldn't be stored in anything other than a secret, for multiple reasons. So you would end up leaking the secret via the diff, but also potentially inside your cluster to other actors who have restricted access.

**Reply by ryanmt** [reply]
I agree completely that this shouldn't be used in a way that exposes secrets to an external system, but IMO that's not intrinsic to a decision to supporting the requested feature by providing a way to generate the diff from the underlying helm/kustomize resource.  

The ability to generate a diff while still using flux is valuable whether or not a user *can* expose secrets by using it.  Especially since exposing secrets is always possible whether or not this is a feature.

**Reply by bburky** [reply]
Argo CD's implementation of diff shows the diff of the templated manifests (not the input Helm values or Kustomize files). When it shows Secret resources it masks all the values inside the secret, but does highlight that the values differ.

**Reply by shibumi** [reply]
Hi, sorry for reviving this thread. I think the mistake is already to have plaintext secrets in any YAML configuration. As far as I know people use Hashicorp Vault or Bitnami's Sealed-Secrets Operator for running around this issue. So flux should always show the diff and any plaintext secrets should not be visible, because SealedSecrets or Vault should handle the rest. Everything else is a cluster misconfiguration IMHO.

**Reply by william-salt-cko** [reply]
@stefanprodan Have you had any further thoughts or discussions about this?

I am currently doing a flux1 -> flux2 migration and lack of diff support has created a quite complex shell script that diffs just helm releases, which doesn't work quite in the same way as the controller. 
We have also on one occasion changed some kustomize and affected a dozen or so releases in production which isn't obvious just from a git diff, whereas a kubectl diff would indicate actually what would happen. 

We have other ideas for workarounds, but we would really like some functionality akin to TF, ArgoCD, or Helm in Flux2, or at least indication there is a plan for it. 🙂

**Reply by RichiCoder1** [reply]
This is something I love about both `terraform` and `ArgoCD`, and it's a bit painful having to maintain a massive diff script just to accomplish this for PRs

**Reply by william-salt-cko** [reply]
@stefanprodan Do you think you could revisit this now? [The PR](https://github.com/kubernetes/kubernetes/pull/96084) for kubectl diff hiding secrets support has been merged in from the original post. Helm diff already supports this too...

**Reply by morremeyer** [reply]
I think we are now in a state where we can take a look at how to do this. The PR is merged as @william-salt-cko mentioned, so this will be part of Kubernetes 1.22.

I started https://github.com/fluxcd/flux2/discussions/820#discussioncomment-821033 to gather ideas.

**Comment by messiahUA** [comment]
I think that diff functionality is a must for production use. For example, I've made a mistake in configuration files and flux completely deleted everything including itself with prune enabled. Of course I can disable pruning, but in this case I would need to do it for everything and it defeats the purpose. These mistakes can be prevented only by checking everything before applying...

**Comment by morremeyer** [comment]
kubectl will have support for hiding secrets in 1.22, see [this PR](https://github.com/kubernetes/kubernetes/pull/96084). 
I think we could start discussing how this can be implemented to support different git flows, so here is something that I would find useful to have:

In most git flows, there is a production branch (most often named `production` or `main`) that flux be configured to apply resources from. If functionality was added that allows the configuration of a staging branch to diff against (e.g. `staging` or `pre-releaes`), this would allow to:

* Use gitflow and gitlabflow natively
* Not add any „blocking“ logic to the deployment from the production branch as the diff and the apply workflows would/could be completely seperate from one another

I have not yet found an idea about how to report the output that I am really happy with. My only idea so far would be to add a URL field to specify a webhook to which the diff could then be posted.

**Reply by morremeyer** [reply]
Another idea I had is for flux to track all branches of the repository and diff changes - this would also enable the usage of githubflow, but probably make the implementation more complex.

However, it would probably also enable easier tracking of diffs by being able to track changes back to a PR (or MR for e.g. GitLab).

**Comment by stefanprodan** [comment]
In Flux [0.26](https://github.com/fluxcd/flux2/releases/tag/v0.26.0) we've introduced the `flux diff kustomization --path my/local/overlay` command which uses the same server-side apply engine as kustomize-controller. The `flux diff` command detects SOPS encrypted secrets and masks the secret values in the diff output.

**Reply by dudicoco** [reply]
@stefanprodan are there any plans to support helm diff as well?

**Reply by stefanprodan** [reply]
Yes. We are currently refactoring helm-controller to allow running some of its logic in Flux CLI.

**Reply by dudicoco** [reply]
Thanks @stefanprodan.

Is there a dedicated issue or a timeline?

**Reply by stefanprodan** [reply]
You can subscribe to this [issue](https://github.com/fluxcd/helm-controller/issues/487) which is a prerequisite for `flux diff helm`.

**Reply by mckennajones** [reply]
Hey there @stefanprodan is the above linked issue in helm-controller still a prereq for implementing a `flux diff hr` command?

**Reply by ghobadimhd** [reply]
Hi, is there any further update about flux diff HelmRelease? Is this feature going to be added to flux or not? I've seen that the above issue has been completed
@stefanprodan

**Reply by onedr0p** [reply]
The helm controller just got a pretty big update so hopefully we're not too far away from this functionality.

**Reply by TaylorChristie** [reply]
Wondering if there is still any interest on the `flux diff helmreleases` command? Would be quite nice to have native support as currently I have a hacky bash script that doesn't quite get me all the way there. Having output similar to `helm diff` with the https://github.com/databus23/helm-diff plugin would be ideal

**Reply by mo-rieger** [reply]
Are there any blockers left that prevent implementing `flux diff helmrelease` command? I am also looking for a better alternative to our bash script.

**Comment by julienFlowdesk** [comment]
The `flux diff kustomization --path X` is an amazing command. I just wanted to add some early feedbacks:

In my case, I am trying to do flux diffing within a github-action with a specific service account having only read-only access to all kubernetes objects within my cluster to limit risks.

I was expecting flux to be able to do diff by reading only the current cluster objects but in fact it looks like it need more than that:

```
.  Kustomization diffing...: running dry-run
 errors occurred:
	* Namespace/x dry-run failed, reason: Forbidden, error: namespaces "x" is forbidden: User "github-action@xxxx" cannot patch resource "namespaces" in API group "" in the namespace "x": requires one of ["container.namespaces.update"] permission(s).

```

Maybe it is the awaited behaviour but I feel like the diff command shouldn't require more than the view roles within kubernetes to output something.

**Reply by stefanprodan** [reply]
Please raise this with the Kubernetes team, Flux performs a server-side apply dry-run, why dry-run requires write access I have no clue.

**Reply by julienFlowdesk** [reply]
For posterity purposes: the kubernetes related issues :

https://github.com/kubernetes/kubernetes/issues/95449
https://github.com/kubernetes/kubectl/issues/981

**Comment by Preisschild** [comment]
Is there any way to show objects that would be created in a new `Kustomization`?

Currently, it just shows `► Kustomization/flux-system/<name> created`

**Reply by stefanprodan** [reply]
Running `flux build ks` will show all objects, and you can use `yq` to selector the one from the diff.
